File: ios/ReservationApp/ReservationApp/Data/Repositories/AuthRepository.swift
//
//  AuthRepository.swift
//  ReservationApp
//
//  Created by Mehmet Ali Kısacık on 13.11.2025.
//

import Foundation

class AuthRepository: AuthRepositoryProtocol {
    private let apiClient: APIClient
    private let keychainManager: KeychainManager
    
    init(apiClient: APIClient = APIClient.shared, keychainManager: KeychainManager = KeychainManager.shared) {
        self.apiClient = apiClient
        self.keychainManager = keychainManager
    }
    
    func login(email: String, password: String) async throws -> AuthResponse {
        let request = LoginRequest(email: email, password: password)
        let response: AuthResponse = try await apiClient.request(
            endpoint: .login,
            body: request,
            responseType: AuthResponse.self
        )
        
        // Save token to keychain
        if !response.authToken.isEmpty {
            _ = keychainManager.saveToken(response.authToken)
        }
        
        return response
    }
    
    func register(request: RegisterRequest) async throws -> AuthResponse {
        let response: AuthResponse = try await apiClient.request(
            endpoint: .register,
            body: request,
            responseType: AuthResponse.self
        )
        
        // Save token to keychain if provided
        if !response.authToken.isEmpty {
            _ = keychainManager.saveToken(response.authToken)
        }
        
        return response
    }
    
    func getCurrentUser() async throws -> User {
        let user: User = try await apiClient.request(
            endpoint: .currentUser,
            responseType: User.self
        )
        return user
    }
}

File: android/domain/src/main/java/com/reservationapp/domain/repository/AuthRepository.kt
package com.reservationapp.domain.repository

import com.reservationapp.core.common.Result
import com.reservationapp.domain.model.AuthResponse
import com.reservationapp.domain.model.RegisterRequest
import com.reservationapp.domain.model.UpdateUserProfileRequest
import com.reservationapp.domain.model.User

interface AuthRepository {
    suspend fun login(email: String, password: String): Result<AuthResponse>
    suspend fun register(request: RegisterRequest): Result<AuthResponse>
    suspend fun getCurrentUser(): Result<User>
    suspend fun updateProfile(request: UpdateUserProfileRequest): Result<User>
    suspend fun logout()
}

